patchwork-extensions-bakedmodel/src/main/java/net/patchworkmc/mixin/extensions/bakedmodel/MixinChunkRenderer.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2020, 2019-2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.patchworkmc.mixin.extensions.bakedmodel;
+
+import java.util.Iterator;
+import java.util.Random;
+import java.util.Set;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.At.Shift;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+import net.minecraftforge.client.model.data.IModelData;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockRenderType;
+import net.minecraft.block.BlockState;
+import net.minecraft.client.render.BufferBuilder;
+import net.minecraft.client.render.RenderLayer;
+import net.minecraft.client.render.block.BlockRenderManager;
+import net.minecraft.client.render.chunk.ChunkOcclusionDataBuilder;
+import net.minecraft.client.render.chunk.ChunkRenderData;
+import net.minecraft.client.render.chunk.ChunkRenderTask;
+import net.minecraft.client.render.chunk.ChunkRenderer;
+import net.minecraft.client.render.chunk.ChunkRendererRegion;
+import net.minecraft.client.render.model.BakedModel;
+import net.minecraft.util.math.BlockPos;
+
+import net.fabricmc.fabric.api.renderer.v1.model.FabricBakedModel;
+import net.fabricmc.fabric.impl.client.indigo.Indigo;
+
+import net.patchworkmc.impl.extensions.bakedmodel.ForgeBlockRenderManager;
+import net.patchworkmc.impl.extensions.bakedmodel.ForgeChunkRenderTask;
+
+/**
+ * Mimic calling Forge's IModelData sensitive version of tesselateBlock().
+ */
+@Mixin(ChunkRenderer.class)
+public abstract class MixinChunkRenderer {
+	/**
+	 * This method can be better if we can talk to the Fabric API developer.
+	 * If the implementation of {@link net.fabricmc.fabric.mixin.client.indigo.renderer.MixinChunkRenderer#hookChunkBuildTesselate}
+	 * is placed in a separate class, it will be much easier for us.
```

Comment:
This is also extremely messy and breaks Canvas/Sodium compat.

Reply:
It is really hard to find a better workaround. The only solution is to talk to the Fabric team to see if they want to make some change to their code to make their hooks less invasive. I don't think we can solve this on our own, wrapping forge's rendering system using fabric's api is unlikely to give good result, because they almost don't share any feature in common.